File: libro/130_fibonacci_2.cpp
/* ********************************************************
* Programa: Fibonacci2
*
* Descripción:
*   Este programa imprime todos los términos
*   de la serie de Fibonacci dentro del rango de
*   valores positivos del tipo lnt: (1 .. INT_MAX)
*   Se imprime tabulando a siete columnas
******************************************************** */

#include <stdio.h>
#include <limits.h>

/*==================
  Variable global
  ==================*/
int TABcolumna;     /* columna a imprimir */
int TABultima;      /* última columna */
int TABancho;       /* ancho de cada columna */


/*=========================================
  Procedimiento para iniciar la tabulación
  con los parámetros indicados
  =========================================*/
void IniciarTabulacion(int columnas, int ancho){
  TABultima  = columnas;
  TABancho   = ancho;
  TABcolumna = 1;
}

/*=======================================*\
  Procedimiento para imprimir tabulando
  a TABultima columnas de TABancho caracteres
\*=======================================*/
void ImprimirTabulando(int k) {
  /* Si se llega al final de las columnas definidas, se pasa a la siguiente fila */
  if (TABcolumna > TABultima) {
    TABcolumna = 1;
    printf( "\n" );
  }
  printf("%*d", TABancho, k);
  TABcolumna++;
}

/*===================*\
  Programa principal
\*===================*/
int main() {
  int termino;    /* término de la serie */
  int anterior;   /* término anterior */
  int aux;

/*-- Iniciar la tabulación --*/
  IniciarTabulacion(6, 11);

/*-- Generar el comienzo de la serie --*/
  anterior = 0;   /* primer número de la serie */
  termino = 1;    /* segundo número de la serie */
  ImprimirTabulando( anterior );
  ImprimirTabulando( termino );

/*-- Generar resto de la serie --*/
  while (INT_MAX - termino >= anterior){
    aux       = termino + anterior ;  /* siguiente número de la serie, suma de los anteriores */
    anterior  = termino;
    termino   = aux;
    ImprimirTabulando( termino );
  }
    printf( "\n" );
}

File: libro/130_fibonacci_2.py
# Programa: Fibonacci2
# Descripción:
#   Este programa imprime todos los términos de la serie de Fibonacci 
#   dentro del rango de valores positivos del tipo lnt: (1 .. INT_MAX)
#   Se imprime tabulando a cuatro columnas

import sys

numeros = [] # Números a imprimir

def Fibonacci(num_1, num_2):

    fibo  = num_1 + num_2
    num_1 = num_2
    num_2 = fibo
    return (fibo, num_1, num_2)


# imprimir tabulando a TABultima columnas de TAB_ancho caracteres
def ImprimirTabulando(numeros, tab_ancho):
    
    texto = ""
    for numero in numeros:    
        texto+=("{num:{fill}{width}d}".format(num=numero, width=tab_ancho, fill= " "))
    print(texto)



# Programa principal
if __name__ == "__main__":

    TAB_columna = 1     # columna a imprimir 
    TAB_columnas = 6    # número de columnas a imprimir 
    TAB_ancho = 11      # ancho de cada columna 
    N = sys.maxsize     # Rango de números a analizar
    anterior = 0        # Número "n"
    termino = 1         # # Número "n+Fibonacci"
    fibo = 0            # Número de Fibonacci calculado con números anterior y término
    numeros = []        # Números a imprimir

    numeros.append(0)   # Primeros números de la serie
    numeros.append(1)
    TAB_columna +=2

    # Serie de Fibonacci
    while(N - termino >= anterior):
        fibo, anterior, termino = Fibonacci(anterior, termino)
        
        #ImprimirTabulando(fibo, TAB_ancho)
        numeros.append(fibo)

        TAB_columna += 1
        # Los números se imprimen en "TAB_ULTIMA" columnas
        # ImprimirTabulando(numeros)      
        if TAB_columna>TAB_columnas:
            TAB_columna = 1
            ImprimirTabulando(numeros, TAB_ancho)
            numeros = []
    
    # Imprimir últimos números añadidos a la serie
    if len(numeros)>0:
        ImprimirTabulando(numeros, TAB_ancho)
        
